File: algorithms/0538.convert-bst-to-greater-tree/solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {TreeNode}
def convert_bst(root)
  @sum = 0
  build_greater_tree(root)
end

def build_greater_tree(node)
  return nil if node.nil?

  build_greater_tree(node.right)

  @sum += node.val
  node.val = @sum

  build_greater_tree(node.left)

  node
end

File: algorithms/0538.convert-bst-to-greater-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func convertBST(root *TreeNode) *TreeNode {
	sum := 0
	helper(root, &sum)

	return root
}

func helper(node *TreeNode, sum *int) {
	if node == nil {
		return
	}

	helper(node.Right, sum)

	*sum += node.Val
	node.Val = *sum

	helper(node.Left, sum)
}
